src/hdmf/query.py

```diff
@@ -160,3 +161,58 @@ def __next__(self):
 
     def next(self):
         return self.dataset.next()
+
+
+class ReferenceResolver(with_metaclass(ABCMeta, object)):
+    """
+    A base class for classes that resolve references
+    """
+
+    @classmethod
+    @abstractmethod
+    def get_inverse_class(cls):
+        """
+        Return the class the represents the ReferenceResolver
+        that resolves refernces to the opposite type.
+
+        BuilderResolver.get_inverse_class should return a class
+        that subclasses ContainerResolver.
+
+        ContainerResolver.get_inverse_class should return a class
+        that subclasses BuilderResolver.
+        """
+        pass
+
+    @abstractmethod
+    def invert(self):
+        """
+        Return an object that defers reference resolution
+        but in the opposite direction.
+        """
+        pass
+
+
+class BuilderResolver(ReferenceResolver):
+    """
+    A reference resolver that resolves references to Builders
+
+    Subclasses should implement invert and the calssmethod get_inverse_class
+
+    BuilderResolver.get_inverse_class should return a class
+    that subclasses ContainerResolver.
+    """
+
+    pass
+
+
+class ContainerResolver(ReferenceResolver):
+    """
+    A reference resolver that resolves references to Containers
+
+    Subclasses should implement invert and the calssmethod get_inverse_class
```

Comment:
Spelling error: "calssmethod " --> "classmethod". The same is in the docstring of BuilderResolver and ContainerResolver

Reply:
Shouldn't subtypes also implement the invert function? If so, this should be mentioned in the docstring as well.